### ISSUE TITLE
Some jobs have a "finished_at" date, but are still started or waiting

### ISSUE THREAD

**Issue by severo** [huggingface/datasets-server#1467]
```
db.jobsBlue.count({"finished_at": {"$exists": true}, "status": {"$nin": ["success", "error", "cancelled"]}})
24
```

For example:

```
{ _id: ObjectId("649f417b849c36335817cfa7"),
  type: 'dataset-size',
  dataset: 'knowrohit07/know_cot',
  revision: 'f89e138e31115fd5b144aa0c52888316e710f752',
  unicity_id: 'dataset-size,knowrohit07/know_cot',
  namespace: 'knowrohit07',
  priority: 'normal',
  status: 'started',
  created_at: 2023-06-30T20:56:27.051Z,
  started_at: 2023-06-30T20:56:27.432Z,
  finished_at: 2023-06-30T20:56:27.499Z }
```

And the cache entry:

```
{
  "_id": { "$oid": "649f4175bd1b024a84d4b8c7" },
  "config": null,
  "dataset": "knowrohit07/know_cot",
  "kind": "dataset-size",
  "split": null,
  "content": {
    "size": {
      "dataset": {
        "dataset": "knowrohit07/know_cot",
        "num_bytes_original_files": 37076410,
        "num_bytes_parquet_files": 13932763,
        "num_bytes_memory": 31107926,
        "num_rows": 74771
      },
      "configs": [
        {
          "dataset": "knowrohit07/know_cot",
          "config": "knowrohit07--know_cot",
          "num_bytes_original_files": 37076410,
          "num_bytes_parquet_files": 13932763,
          "num_bytes_memory": 31107926,
          "num_rows": 74771,
          "num_columns": 3
        }
      ],
      "splits": [
        {
          "dataset": "knowrohit07/know_cot",
          "config": "knowrohit07--know_cot",
          "split": "train",
          "num_bytes_parquet_files": 13932763,
          "num_bytes_memory": 31107926,
          "num_rows": 74771,
          "num_columns": 3
        }
      ]
    },
    "pending": [],
    "failed": []
  },
  "dataset_git_revision": "f89e138e31115fd5b144aa0c52888316e710f752",
  "details": null,
  "error_code": null,
  "http_status": 200,
  "job_runner_version": 2,
  "progress": 1,
  "updated_at": { "$date": "2023-06-30T20:56:27.488Z" }
}
```

**Comment by severo** [comment]
Note that these jobs are never deleted (the TTL index only deletes finished jobs with one of the three valid finished status)

**Comment by severo** [comment]
Run today: we have 20 jobs with status "started", and one with status "finished" (this status does not exist in the code!)

**Comment by severo** [comment]
I deleted them manually

**Comment by severo** [comment]
<img width="363" alt="Capture d’écran 2023-07-26 à 15 20 01" src="https://github.com/huggingface/datasets-server/assets/1676121/eef31f27-b45c-47e1-be32-25bdf2931d7e">

**Comment by severo** [comment]
Today:

```
> db.jobsBlue.count({"finished_at": {"$exists": true}, "status": {"$nin": ["success", "error", "cancelled"]}})
< 16
```

```
> db.jobsBlue.find({"finished_at": {"$exists": true}, "status": {"$nin": ["success", "error", "cancelled"]}}, {"status": 1, "_id": 0})
< { status: 'started' } x16
```

**Comment by lhoestq** [comment]
It seems to happen only for super fast job. Maybe a race condition between setting the job as "started" and as "finished" ?

**Comment by lhoestq** [comment]
Actually I find that the timestamps don't make sense. e.g. for Salesforce/dialogstudio

for the Job:
- created_at = 2023-08-16T16:34:40.864000
- started_at = 2023-08-16T16:34:41.264000
- finished_at = 2023-08-16T16:34:41.282000

but for the Cache
- updated_at = 2023-08-16T16:34:39.983+00:00

which is **before** the job ran O_O

**Comment by lhoestq** [comment]
Alright it turns out this is because multiple workers process the same job at the same time (up to 5 concurrent duplicates)

**Comment by severo** [comment]
Seems to be fixed by #1699

**Comment by AndreaFrancis** [comment]
I found a new started + finished_at job in the queue:
![Screenshot from 2023-08-18 16-57-40](https://github.com/huggingface/datasets-server/assets/5564745/4bdfb9c6-06c2-4d48-b295-8e48e250465d)


Should I reopen this issue?

**Comment by AndreaFrancis** [comment]
Today:
```
db.jobsBlue.countDocuments({status:"started", finished_at:{$exists:true}})
10

```

I will remove them to avoid blocking the queue

**Comment by lhoestq** [comment]
Ok I'm taking a look 👀

**Comment by AndreaFrancis** [comment]
We could change the logic of the next waiting job.

```

 JobDocument.objects(
                status=Status.WAITING, namespace__nin=set(started_job_namespaces), priority=priority, **filters
            )
            .order_by("+created_at")
            .only("type", "dataset", "revision", "config", "split", "priority", "unicity_id")
            .no_cache()
            .first()


```
Instead of ordering by created_at to get the first one, maybe just choose one random record so that it will reduce the probability of processing the same job.
I know that is not the idea we want about a priority queue, but at least we would be increasing the number of processed jobs.

**Comment by AndreaFrancis** [comment]
I found a couple of interesting documentation/concepts about AWS SQS that maybe can give us some ideas how to handle the issues with the queue:
Context:   https://tomgregory.com/3-surprising-facts-about-aws-sqs-fifo-queues/ 

**- FIFO logic** -> SQS by default uses time as priority logic,  we use it when getting the next pending job ordered by `created_at` field
**- message_group_id** -> it is possible to tag every message with a "message_group_id" that is like a partition, SQS limits to deliver only one message per "message_group_id", same as we do to when getting the next job (with a namespace different to any of the already being processed)

_Messages that have the same messages group id will be returned in order. While this is intended behaviour for a FIFO queue, remember that only once a message has been removed from the queue will the next message with the same message group id be returned._

**- Visibility timeout** -> This is exactly the logic we need to "lock read" by job runners!, that is, once a worker would read a message, we can set the time the message won't be available to read by another worker. It could be the approximate time we think a job runner finishes a job.

_Immediately after a message is received, it remains in the queue. To prevent other consumers from processing the message again, Amazon SQS sets a visibility timeout, a period of time during which Amazon SQS prevents all consumers from receiving and processing the message. The default visibility timeout for a message is 30 seconds. The minimum is 0 seconds. The maximum is 12 hours._ 
From: https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-visibility-timeout.html

- It is possible to count the number of records in a queue -> Used for metrics
- Is it possible to delete a message from the queue? (We need it when canceling a job) 

Maybe it is not a priority to change the stack we currently use for the queue, but maybe can give us ideas for future implementations.

**Comment by severo** [comment]
Thanks for the research! Indeed, MongoDB is currently the bottleneck which prevents us to scale the workers, both for the jobs, when it's full, and for the cache in some cases. We have to change something on the jobs side, because we have issues to often. Maybe switching to SQS or something similar installed on our infra?

**Comment by lhoestq** [comment]
No new case so far - let's wait a bit more before closing

**Comment by lhoestq** [comment]
still no new case - closing (hopefully for good)
